src/main/java/edu/kit/compiler/optimizations/unrolling/LoopAnalysis.java

```diff
@@ -0,0 +1,316 @@
+package edu.kit.compiler.optimizations.unrolling;
+
+import static firm.bindings.binding_irgraph.ir_graph_properties_t.IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import edu.kit.compiler.io.StackWorklist;
+import edu.kit.compiler.optimizations.Util;
+import firm.BlockWalker;
+import firm.Graph;
+import firm.bindings.binding_irdom;
+import firm.bindings.binding_irnode.ir_opcode;
+import firm.nodes.Block;
+import firm.nodes.Cond;
+import firm.nodes.Node;
+import firm.nodes.Proj;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+/**
+ * Implements an analysis to detect loops in Firm graphs.
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+public final class LoopAnalysis {
```

Comment:
Do we need an additional loop analysis?

Reply:
You're right that we should ideally merge the two analyses. However, given the time frame, I think it is of secondary importance. Our respective loop analyses don't necessarily compute the same set of information, so combining them would not be trivial.